docs: VP-19 Terraform provider publishing ADR (public/private) + Terraform provider parity spec (#12)
mikecbrant/verified-permissions-authorizer: docs/adr-0002-terraform-provider-publishing.md

# ADR 0002: Terraform provider publishing options (public and private)

> Status: Proposed

Reviewer: ```suggestion
Status: Accepted
```

Date: 2025-09-10

Decision drivers
- Ship a Terraform provider that mirrors the Pulumi provider’s functionality and configuration so teams on Terraform can adopt the same authorizer pattern.
- Choose a publishing model that balances discoverability, consumer friction, signing/security, and our ability to maintain releases.
- Keep the build/release pipeline simple and automatable from this repo.

## Summary

> We need to distribute a Terraform provider that provisions the Verified Permissions policy store, deploys the bundled Lambda authorizer, optionally provisions Cognito (with SES email), and ingests schema/policies with the same validations as the Pulumi provider. This ADR evaluates publishing options—public and private—and recommends a path.

Reviewer: ```suggestion
We need to distribute a Terraform provider that provisions the Verified Permissions policy store, deploys the bundled Lambda authorizer, optionally provisions Cognito, and ingests schema/policies with the same validations as the Pulumi provider. This ADR evaluates publishing options—public and private—and recommends a path.
```

## In-scope
- Publishing channels for the Terraform provider and their implications
- Ownership/namespace, access control, signing, and release processes
- Artifact format and pipeline requirements

Out of scope
- Implementation details of the provider (covered by the provider spec in `docs/vp-19-terraform-provider-spec.md`)

## Options

### A. Public: Terraform Registry (registry.terraform.io)

- What it is: The canonical public distribution for providers. Provider appears as `NAMESPACE/NAME` in the public registry.

> - Ownership/namespace: A GitHub organization or user namespace must be linked to a Terraform Registry publisher account; provider repository typically named `terraform-provider-<name>`. We must select the namespace (e.g., `mikecbrant` or an org such as `charlielabs`).

Reviewer: The namespace for publicly published provider should be `mikecbrant`.

- Access control: Public read access; releases are driven by signed GitHub releases in the repo.
- Discoverability: Highest (searchable on the public registry, shown in `terraform init` messages).
- Consumer friction: Lowest. Standard `required_providers` with `source = "namespace/name"` and `version` constraint. No extra CLI config.
- Security/signing: Registry requires a checksums file and a signature for each release. Typical pipeline uses GPG to sign the `SHA256SUMS` file (or uses the registry-supported signing method). Keys must be managed and rotated if needed. Terraform CLI verifies provider signatures on install. See HashiCorp docs for publishing and signature verification details. [Citations below]

> - Versioning: SemVer tags (e.g., `v0.1.0`). Registry tracks versions and deprecations.

Reviewer: I would like to use our changesets functionality to increment both this and the Pulumi provider.   Ideally I would like versions to always exactly match for both providers such that TF provider X.Y.Z has same functionality as Pulumi provider X.Y.Z

Please update .charlie documentation files as appropriate to enforce this approach.

- Organizational prerequisites: Shared storage and a documented setup guide.

When to choose: Temporary internal distribution or bootstrap for air‑gapped teams.

## Comparison

- Discoverability: A (highest) > B > C/D
- Consumer friction: A/B (lowest) < C < D
- Security and signing: All require robust signing; A/B provide built‑in verification UX.
- Docs UX: A/B render provider docs; C/D require a separate docs site.
- Ongoing cost: A/B lowest (managed), C/D higher (operate and support).

## Recommendation
Choose Option A now: publish publicly to the Terraform Registry under a project‑controlled namespace.

Rationale
- Minimal friction for adopters and parity with how many providers are consumed today.
- We already maintain source in a public repo; publishing adds little incremental risk.
- The same signed artifacts can be used later for a private mirror if needed.

If organizational requirements dictate private distribution later, Option B (Private Registry in HCP Terraform) is the preferred private path due to low friction and built‑in docs/UX. Options C/D remain viable for air‑gapped deployments.

## Build & Release implications (for Options A/B)

> - Repository layout: `terraform-provider-<name>` (or this repo with a `providers/terraform` subdir; see open questions).

Reviewer: Change to reflect use of subdir in this repo.  We would like to use the same exact Go code for both providers to whatever extent possible so the provider logic itself might need to be moved into new directory, with terraform and pulumi subdirs only holding logic for provider-specific inputs and output (exports) mappings.

- Artifacts per release:
  - `terraform-provider-<name>_vX.Y.Z_<os>_<arch>.zip` for supported platforms (Linux/amd64, Linux/arm64, Darwin/amd64, Darwin/arm64, Windows/amd64 at minimum).
  - `SHA256SUMS` and a detached signature (e.g., `SHA256SUMS.sig`).
  - Provider manifest (generated by tooling) that declares supported protocol versions.
- CI steps:

>   1) Build with Go `-trimpath` and reproducible flags.

Reviewer: We need to ensure that Pulumi publish and TF publish operate in parallel but that there is only single common workflow step after successful publish in both locations that updates the repo with final version information commit.  Both of these publishing steps should also use comment changesets-based release determination step.

  2) Archive per platform with correct naming.
  3) Generate checksums and sign them with the registry‑recognized method.
  4) Create a GitHub Release for tag `vX.Y.Z` including the zips, checksums, and signature.
  5) Registry picks up the release and serves it to `terraform init`.

## Organizational/account prerequisites
- Select and control the namespace: GitHub org/user that will own the provider and the registry publisher.
- Create/secure a signing key for releases (stored as CI secret; establish rotation plan).
- Reserve provider name and initialize the repository (either a new repo named `terraform-provider-<name>` or a subdirectory in this monorepo with Release workflow uploading from this repo).

## Assumptions

> - The provider will be implemented in Go using the official Terraform Plugin Framework (protocol v6) unless we decide otherwise during spec review.

Reviewer: My preference would be to re-use as much of the existing provider logic as possible so that it is the same exact logic for both providers.

- We will target Linux (amd64/arm64), macOS (amd64/arm64), and Windows (amd64) initially.

> - This repo remains the single source of truth for the Lambda code and the Pulumi provider; the Terraform provider can live alongside them unless we decide to split repos.

Reviewer: As noted. We want everything in this repo, with re-use of Go logic to as great of extent as possible

  5) For Pulumi: publish component provider/SDK packages matching the same `vX.Y.Z`.

Failure handling
- Treat both publishes as a unit. If either publish fails, do not run the common version‑update step; investigate, fix, and retry the failed job so both providers publish the same version.

### Versioning
- Providers must maintain strict 1:1 version parity. The Pulumi and Terraform providers are versioned and released together with identical SemVer (`X.Y.Z`) numbers, coordinated by the same Changeset and automated pipelines.

## Organizational/account prerequisites
- Select and control the namespace: GitHub org/user that will own the provider and the registry publisher.
- Create/secure a signing key for releases (stored as CI secret; establish rotation plan).
- Reserve provider name and initialize the repository subdirectory (`providers/terraform`) in this monorepo with Release workflow uploading from this repo.
- Internal documentation under `.charlie/` encodes the synchronized versioning and release workflow requirements (Changesets usage, parallel Pulumi/Terraform publish, and the common post‑publish version update step).

## Assumptions

> - The provider will be implemented in Go using the official Terraform Plugin Framework (protocol v6) unless we decide otherwise during spec review.

Reviewer: ```suggestion
- The provider will be implemented in Go using common logic to the terraform provider.
```
